File: sdk/python/polar/meters.py
from __future__ import annotations

import builtins
import typing

from polar.base import AsyncServiceBase, SyncServiceBase, parse_response
from polar.errors import (
    HTTPValidationError,
    ResourceNotFound,
)
from polar.inputs import (
    MetadataQuery,
    MeterCreate,
    MeterUpdate,
)
from polar.literals import (
    AggregationFunction,
    MeterSortProperty,
    TimeInterval,
    Timezone,
)
from polar.outputs import (
    ListResourceMeter,
    Meter,
    MeterQuantities,
)


class MetersSync(SyncServiceBase):
    def list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        query: str | None = None,
        is_archived: bool | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[MeterSortProperty] | None = ["name"],
        metadata: MetadataQuery = None,
    ) -> ListResourceMeter:
        """
        List meters.

        **Scopes**: `meters:read` `meters:write`

        Args:
            organization_id: Filter by organization ID.
            query: Filter by name.
            is_archived: Filter on archived meters.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.
            metadata: Filter by metadata key-value pairs. It uses the `deepObject` style, e.g. `?metadata[key]=value`.

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/meters/",
            path_params={},
            query_params={
                "organization_id": organization_id,
                "query": query,
                "is_archived": is_archived,
                "page": page,
                "limit": limit,
                "sorting": sorting,
                "metadata": metadata,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(response, ListResourceMeter, method_errors)

    def create(
        self,
        **kwargs: typing.Unpack[MeterCreate],
    ) -> Meter:
        """
        Create a meter.

        **Scopes**: `meters:write`

        Args:

        Raises:
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/meters/",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(response, Meter, method_errors)

    def get(
        self,
        id: str,
    ) -> Meter:
        """
        Get a meter by ID.

        **Scopes**: `meters:read` `meters:write`

        Args:
            id: The meter ID.

        Raises:
            ResourceNotFound: Meter not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/meters/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, Meter, method_errors)

    def update(
        self,
        id: str,
        **kwargs: typing.Unpack[MeterUpdate],
    ) -> Meter:
        """
        Update a meter.

        **Scopes**: `meters:write`

        Args:
            id: The meter ID.

        Raises:
            ResourceNotFound: Meter not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/meters/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, Meter, method_errors)

    def quantities(
        self,
        id: str,
        *,
        start_timestamp: str,
        end_timestamp: str,
        interval: TimeInterval,
        timezone: Timezone = "UTC",
        customer_id: str | builtins.list[str] | None = None,
        external_customer_id: str | builtins.list[str] | None = None,
        customer_aggregation_function: AggregationFunction | None = None,
        metadata: MetadataQuery = None,
    ) -> MeterQuantities:
        """
        Get quantities of a meter over a time period.

        **Scopes**: `meters:read` `meters:write`

        Args:
            id: The meter ID.
            start_timestamp: Start timestamp.
            end_timestamp: End timestamp.
            interval: Interval between two timestamps.
            timezone: Timezone to use for the timestamps. Default is UTC.
            customer_id: Filter by customer ID.
            external_customer_id: Filter by external customer ID.
            customer_aggregation_function: If set, will first compute the quantities per customer before aggregating them using the given function. If not set, the quantities will be aggregated across all events.
            metadata: Filter by metadata key-value pairs. It uses the `deepObject` style, e.g. `?metadata[key]=value`.

        Raises:
            ResourceNotFound: Meter not found.
            HTTPValidationError: Validation Error
            PolarNetworkError: Raised when a network error occurs while making the request.
            PolarServerError: Raised when the server returns a 5xx error response.
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/meters/{id}/quantities",
            path_params={
                "id": id,
            },
            query_params={
                "start_timestamp": start_timestamp,
                "end_timestamp": end_timestamp,
                "interval": interval,
                "timezone": timezone,
                "customer_id": customer_id,
                "external_customer_id": external_customer_id,
                "customer_aggregation_function": customer_aggregation_function,
                "metadata": metadata,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, MeterQuantities, method_errors)


class MetersAsync(AsyncServiceBase):
    async def list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        query: str | None = None,
        is_archived: bool | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[MeterSortProperty] | None = ["name"],
        metadata: MetadataQuery = None,
    ) -> ListResourceMeter:
        """
        List meters.

        **Scopes**: `meters:read` `meters:write`

        Args:
            organization_id: Filter by organization ID.
            query: Filter by name.
            is_archived: Filter on archived meters.
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.
            metadata: Filter by metadata key-value pairs. It uses the `deepObject` style, e.g. `?metadata[key]=value`.

        Raises:
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/meters/",
            path_params={},
            query_params={
                "organization_id": organization_id,
                "query": query,
                "is_archived": is_archived,
                "page": page,
                "limit": limit,
                "sorting": sorting,
                "metadata": metadata,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(response, ListResourceMeter, method_errors)

    async def create(
        self,
        **kwargs: typing.Unpack[MeterCreate],
    ) -> Meter:
        """
        Create a meter.

        **Scopes**: `meters:write`

        Args:

        Raises:
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="POST",
            url="/v1/meters/",
            path_params={},
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(response, Meter, method_errors)

    async def get(
        self,
        id: str,
    ) -> Meter:
        """
        Get a meter by ID.

        **Scopes**: `meters:read` `meters:write`

        Args:
            id: The meter ID.

        Raises:
            ResourceNotFound: Meter not found.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/meters/{id}",
            path_params={
                "id": id,
            },
            query_params={},
        )
        response = await self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, Meter, method_errors)

    async def update(
        self,
        id: str,
        **kwargs: typing.Unpack[MeterUpdate],
    ) -> Meter:
        """
        Update a meter.

        **Scopes**: `meters:write`

        Args:
            id: The meter ID.

        Raises:
            ResourceNotFound: Meter not found.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/meters/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, Meter, method_errors)

    async def quantities(
        self,
        id: str,
        *,
        start_timestamp: str,
        end_timestamp: str,
        interval: TimeInterval,
        timezone: Timezone = "UTC",
        customer_id: str | builtins.list[str] | None = None,
        external_customer_id: str | builtins.list[str] | None = None,
        customer_aggregation_function: AggregationFunction | None = None,
        metadata: MetadataQuery = None,
    ) -> MeterQuantities:
        """
        Get quantities of a meter over a time period.

        **Scopes**: `meters:read` `meters:write`

        Args:
            id: The meter ID.
            start_timestamp: Start timestamp.
            end_timestamp: End timestamp.
            interval: Interval between two timestamps.
            timezone: Timezone to use for the timestamps. Default is UTC.
            customer_id: Filter by customer ID.
            external_customer_id: Filter by external customer ID.
            customer_aggregation_function: If set, will first compute the quantities per customer before aggregating them using the given function. If not set, the quantities will be aggregated across all events.
            metadata: Filter by metadata key-value pairs. It uses the `deepObject` style, e.g. `?metadata[key]=value`.

        Raises:
            ResourceNotFound: Meter not found.
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/meters/{id}/quantities",
            path_params={
                "id": id,
            },
            query_params={
                "start_timestamp": start_timestamp,
                "end_timestamp": end_timestamp,
                "interval": interval,
                "timezone": timezone,
                "customer_id": customer_id,
                "external_customer_id": external_customer_id,
                "customer_aggregation_function": customer_aggregation_function,
                "metadata": metadata,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            404: ResourceNotFound,
            422: HTTPValidationError,
        }
        return parse_response(response, MeterQuantities, method_errors)

File: sdk/typescript/src/services/meters.ts
import { ClientBase } from "../base";
import type { MetadataQuery, MeterCreate, MeterUpdate } from "../models/inputs";
import type { ListResourceMeter, Meter, MeterQuantities } from "../models/outputs";
import type {
  AggregationFunction,
  MeterSortProperty,
  TimeInterval,
  Timezone,
} from "../models/literals";
import { HTTPValidationError, ResourceNotFound } from "../errors";

export const listMeters = (client: ClientBase) => {
  /**
   * List meters.
   *
   * **Scopes**: `meters:read` `meters:write`
   *
   * @param query - Query parameters
   * @returns {ListResourceMeter}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (query?: {
    organization_id?: string | string[] | null;
    query?: string | null;
    is_archived?: boolean | null;
    page?: number;
    limit?: number;
    sorting?: MeterSortProperty[] | null;
    metadata?: MetadataQuery;
  }): Promise<ListResourceMeter> => {
    const pathParams = {};
    const queryParams = {
      organization_id: query?.organization_id,
      query: query?.query,
      is_archived: query?.is_archived,
      page: query?.page ?? 1,
      limit: query?.limit ?? 10,
      sorting: query?.sorting ?? ["name"],
      metadata: query?.metadata,
    };
    const request = client.buildRequest("GET", "/v1/meters/", pathParams, queryParams, undefined);
    const response = await client.sendRequest(request);
    return client.parseResponse<ListResourceMeter>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
/**
 * List meters.
 *
 * **Scopes**: `meters:read` `meters:write`
 *
 * @param query - Query parameters
 * @returns {AsyncGenerator<Meter>} A generator that yields items of type Meter.
 * @throws {PolarNetworkError} When a network error occurs
 * @throws {PolarServerError} When the server returns a 5xx error
 * @throws {HTTPValidationError} Validation Error
 */
export const iterlistMeters = (client: ClientBase) => {
  return async function* (query?: {
    organization_id?: string | string[] | null;
    query?: string | null;
    is_archived?: boolean | null;
    page?: number;
    limit?: number;
    sorting?: MeterSortProperty[] | null;
    metadata?: MetadataQuery;
  }): AsyncGenerator<Meter> {
    let page: number;
    page = query?.page ?? 1;
    let limit: number | undefined;
    limit = query?.limit;

    while (true) {
      const response = await listMeters(client)({ ...query, page, limit });
      for (const item of response.items) {
        yield item;
      }
      if (page >= response.pagination.max_page) {
        break;
      }
      page++;
    }
  };
};
export const createMeters = (client: ClientBase) => {
  /**
   * Create a meter.
   *
   * **Scopes**: `meters:write`
   *
   * @param body - Request body
   * @returns {Meter}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (body: MeterCreate): Promise<Meter> => {
    const pathParams = {};
    const queryParams = {};
    const request = client.buildRequest("POST", "/v1/meters/", pathParams, queryParams, body);
    const response = await client.sendRequest(request);
    return client.parseResponse<Meter>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
export const getMeters = (client: ClientBase) => {
  /**
   * Get a meter by ID.
   *
   * **Scopes**: `meters:read` `meters:write`
   *
   * @param id - The meter ID.
   * @returns {Meter}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Meter not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string): Promise<Meter> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "GET",
      "/v1/meters/{id}",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<Meter>(response, "json", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
export const updateMeters = (client: ClientBase) => {
  /**
   * Update a meter.
   *
   * **Scopes**: `meters:write`
   *
   * @param id - The meter ID.
   * @param body - Request body
   * @returns {Meter}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Meter not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (id: string, body: MeterUpdate): Promise<Meter> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest("PATCH", "/v1/meters/{id}", pathParams, queryParams, body);
    const response = await client.sendRequest(request);
    return client.parseResponse<Meter>(response, "json", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};
export const quantitiesMeters = (client: ClientBase) => {
  /**
   * Get quantities of a meter over a time period.
   *
   * **Scopes**: `meters:read` `meters:write`
   *
   * @param id - The meter ID.
   * @param query - Query parameters
   * @returns {MeterQuantities}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {ResourceNotFound} Meter not found.
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    id: string,
    query: {
      start_timestamp: string;
      end_timestamp: string;
      interval: TimeInterval;
      timezone?: Timezone;
      customer_id?: string | string[] | null;
      external_customer_id?: string | string[] | null;
      customer_aggregation_function?: AggregationFunction | null;
      metadata?: MetadataQuery;
    },
  ): Promise<MeterQuantities> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {
      start_timestamp: query.start_timestamp,
      end_timestamp: query.end_timestamp,
      interval: query.interval,
      timezone: query?.timezone ?? "UTC",
      customer_id: query?.customer_id,
      external_customer_id: query?.external_customer_id,
      customer_aggregation_function: query?.customer_aggregation_function,
      metadata: query?.metadata,
    };
    const request = client.buildRequest(
      "GET",
      "/v1/meters/{id}/quantities",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request);
    return client.parseResponse<MeterQuantities>(response, "json", {
      404: ResourceNotFound,
      422: HTTPValidationError,
    });
  };
};

export function createMetersService(client: ClientBase) {
  return {
    list: listMeters(client),
    create: createMeters(client),
    get: getMeters(client),
    update: updateMeters(client),
    quantities: quantitiesMeters(client),
    iterlist: iterlistMeters(client),
  };
}

export type Meters = ReturnType<typeof createMetersService>;
